src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java

```diff
@@ -0,0 +1,130 @@
+
+  static long getMaximumExistingSequenceNumber(DeltaTargetContext context, String project, @Nullable String datasetName,
+                                               BigQuery bigQuery, EncryptionConfiguration encryptionConfiguration) {
+    StringBuilder builder = new StringBuilder();
+    builder.append("SELECT MAX(max_sequence_num) FROM (");
+    Set<SourceTable> allTables = context.getAllTables();
+    List<String> maxSequenceNumQueryPerTable = new ArrayList<>();
+    for (SourceTable table : allTables) {
+      TableId tableId = TableId.of(project,
+                                   datasetName != null ? normalize(datasetName) : normalize(table.getDatabase()),
+                                   normalize(table.getTable()));
+      if (bigQuery.getTable(tableId) != null) {
+        maxSequenceNumQueryPerTable.add(String.format("SELECT MAX(_sequence_num) as max_sequence_num FROM %s.%s.%s",
+                                                      tableId.getProject(), tableId.getDataset(), tableId.getTable()));
+      }
+    }
+
+    builder.append(String.join(" UNION ", maxSequenceNumQueryPerTable));
+    builder.append(");");
+
+    long maxSequenceNumber;
+    try {
+      maxSequenceNumber = maxSequenceNumQueryPerTable.size() == 0 ? 0 : executeMaxQuery(bigQuery, builder.toString(),
+                                                                                        encryptionConfiguration);
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Failed to compute the maximum sequence number among all existing tables.", e);
+    }
+    return maxSequenceNumber;
+  }
+
+  static long getMaximumSequenceNumberForTable(BigQuery bigQuery, TableId tableId,
+                                               EncryptionConfiguration encryptionConfig) throws Exception {
+    if (bigQuery.getTable(tableId) == null) {
+      return 0L;
+    }
+
+    String query = String.format("SELECT MAX(_sequence_num) FROM %s.%s", tableId.getDataset(), tableId.getTable());
+    return executeMaxQuery(bigQuery, query, encryptionConfig);
+  }
+
+  private static long executeMaxQuery(BigQuery bigQuery, String query, EncryptionConfiguration encryptionConfig)
+    throws InterruptedException {
+    QueryJobConfiguration.Builder jobConfigBuilder = QueryJobConfiguration.newBuilder(query);
+    if (encryptionConfig != null) {
+      jobConfigBuilder.setDestinationEncryptionConfiguration(encryptionConfig);
+    }
+    QueryJobConfiguration jobConfig = jobConfigBuilder.build();
+    JobId jobId = JobId.of(UUID.randomUUID().toString());
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(jobConfig).setJobId(jobId).build());
+    queryJob.waitFor();
+    TableResult result = queryJob.getQueryResults();
+    Iterator<FieldValueList> resultIter = result.iterateAll().iterator();
+    if (!resultIter.hasNext()) {
+      return 0L;
+    }
+    // query is SELECT MAX(_sequence_num) FROM ...
+    // so there is at most one row and one column in the output.
```

Comment:
// the query passed in is expected to generate an aggregated result which contains one row and one column.

Reply:
fixed.